x-pack/plugins/fleet/server/routes/epm/handlers.ts

```diff
@@ -102,22 +106,51 @@ export const getFileHandler: RequestHandler<TypeOf<typeof GetFileRequestSchema.p
 ) => {
   try {
     const { pkgName, pkgVersion, filePath } = request.params;
-    const registryResponse = await getFile(`/package/${pkgName}/${pkgVersion}/${filePath}`);
-
-    const headersToProxy: KnownHeaders[] = ['content-type', 'cache-control'];
-    const proxiedHeaders = headersToProxy.reduce((headers, knownHeader) => {
-      const value = registryResponse.headers.get(knownHeader);
-      if (value !== null) {
-        headers[knownHeader] = value;
+    const savedObjectsClient = context.core.savedObjects.client;
+    const savedObject = await getInstallationObject({ savedObjectsClient, pkgName });
+    const pkgInstallSource = savedObject?.attributes.install_source;
+    // TODO: when package storage is available, remove installSource check and check cache and storage, remove registry call
+    if (pkgInstallSource === 'upload') {
+      const headerContentType = mime.contentType(path.extname(filePath));
+      if (!headerContentType) {
+        return response.custom({
+          body: 'file type is invalid',
+          statusCode: 400,
+        });
       }
-      return headers;
-    }, {} as ResponseHeaders);
+      const archiveFile = getArchiveEntry(`${pkgName}-${pkgVersion}/${filePath}`);
+      if (!archiveFile) {
+        return response.custom({
+          body: 'file does not exist in cache',
```

Comment:
Let's remove mention of a cache.  Also, how about something mentioning it's an uploaded package? Maybe "uploaded package file not found: $filePath"